Meter.c

```diff
@@ -288,12 +288,16 @@ static const char* const GraphMeterMode_dotsAscii[] = {
 
 static void GraphMeterMode_draw(Meter* this, int x, int y, int w) {
    const ProcessList* pl = this->pl;
-
-   if (!this->drawData) {
-      this->drawData = xCalloc(1, sizeof(GraphData));
+   GraphData* data = &this->drawData;
+
+   if ((unsigned int)w * 2 > data->nValues) {
+      unsigned int oldNValues = data->nValues;
+      data->nValues = w * 2;
+      data->values = xReallocArray(data->values, data->nValues, sizeof(*data->values));
+      memmove(data->values + (data->nValues - oldNValues), data->values, oldNValues * sizeof(*data->values));
+      memset(data->values, 0, (data->nValues - oldNValues) * sizeof(*data->values));
```

Comment:
There will be small performance problems if you reallocate buffer this way.

1.  `realloc(3)` internally moves the data from the old buffer to the new one, and you are going shift the graph data to the new location (through `memmove`). This moves the data twice.
2. If `htop` is run on a terminal emulator window, and the user resizes the window with the mouse, there would be a lot of window change events (the terminal width increases one by one for each event), and there would be a a lot of reallocations.

For the second point, I'm not sure if you like the buffer size to be rounded to a power of two.

Reply:
It doesn't hurt over-allocating here. Thus if a resize is needed, then just allocate a bunch of extra entries (like 1/8th of the existing buffer). This will reduce the number of necessary re-allocations by a lot.

The first case internally already checks if there is need to re-allocate, so the double-copying is only there in case of actual size changes for the buffer. I think if this is rare enough we should be able to live with that.

Reply:
Yeah, a multiplicative allocation increase would probably be better; I'll add that.

Reply:
@zevweiss I'm working on another version of the `GraphData` allocation and may file a pull request separately from yours.